lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -193,31 +271,24 @@ static ParseResult parseFIRRTLType(FIRRTLType &result,
   return parseFIRRTLType(result, name, parser);
 }
 
-/// Parse a type registered to this dialect.
-Type FIRRTLDialect::parseType(DialectAsmParser &parser) const {
-  llvm::StringRef name;
-  if (parser.parseKeyword(&name))
-    return Type();
+//===----------------------------------------------------------------------===//
+// Dialect Type Parsing and Printing
+//===----------------------------------------------------------------------===//
 
-  // Parse a generated type.
-  Type genType;
-  auto parseResult = generatedTypeParser(getContext(), parser, name, genType);
-  if (parseResult.hasValue())
-    return genType;
+/// Print a type registered to this dialect.
+void FIRRTLDialect::printType(Type type, DialectAsmPrinter &os) const {
+  type.cast<FIRRTLType>().print(os);
```

Comment:
Isn't this the path that will print `CMemory` types? How is this not crashing?

Reply:
You're totally right, that's a typo. Should use the static printType above.

Reply:
Its weird that it wasn't [asserting](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/Types.h#L245), I was really confused

Reply:
Yeah that surprised me too at first. But I think what's happening is that if you hold a `CMemoryType` directly, it will use the `print` declared on that struct directly, with no dispatch. Only when you funnel through `mlir::Type`'s printing you arrive at this path through the dialect, which I guess wasn't happing that often.